src/Interpreters/getCustomKeyFilterForParallelReplicas.cpp

```diff
@@ -40,52 +40,98 @@ ASTPtr getCustomKeyFilterForParallelReplica(
+                throw Exception(
+                    ErrorCodes::INVALID_SETTING_VALUE,
+                    "Invalid custom key range upper bound: {}. Value must be smaller than custom key column type (UInt16) max value",
+                    range_upper);
+        }
         else if (typeid_cast<const DataTypeUInt8 *>(custom_key_column_type.get()))
-            size_of_universum = RelativeSize(std::numeric_limits<UInt8>::max()) + RelativeSize(1);
+        {
+            range_upper = filter.range_upper > 0 ? RelativeSize(filter.range_upper) + RelativeSize(1)
+                                                 : RelativeSize(std::numeric_limits<UInt8>::max()) + RelativeSize(1);
+            if (range_upper > RelativeSize(std::numeric_limits<UInt8>::max()) + RelativeSize(1))
+                throw Exception(
+                    ErrorCodes::INVALID_SETTING_VALUE,
+                    "Invalid custom key range upper bound: {}. Value must be smaller than custom key column type (UInt8) max value",
+                    range_upper);
+        }
     }
 
-    if (size_of_universum == RelativeSize(0))
+    if (range_upper == RelativeSize(0))
         throw Exception(
             ErrorCodes::ILLEGAL_TYPE_OF_COLUMN_FOR_FILTER,
             "Invalid custom key column type: {}. Must be one unsigned integer type",
             custom_key_column_type->getName());
 
+    if (range_lower >= range_upper)
+        throw Exception(
+            ErrorCodes::INVALID_SETTING_VALUE,
+            "Invalid custom key filter range: Range upper bound {} must be larger than range lower bound {}",
+            range_lower,
+            range_upper);
+
+    RelativeSize size_of_universum = range_upper - range_lower;
+
+    if (size_of_universum <= RelativeSize(replicas_count))
+        throw Exception(
+            ErrorCodes::INVALID_SETTING_VALUE, "Invalid custom key filter range: Range must be larger than than the number of replicas");
+
     RelativeSize relative_range_size = RelativeSize(1) / replicas_count;
     RelativeSize relative_range_offset = relative_range_size * RelativeSize(replica_num);
 
     /// Calculate the half-interval of `[lower, upper)` column values.
     bool has_lower_limit = false;
     bool has_upper_limit = false;
 
-    RelativeSize lower_limit_rational = relative_range_offset * size_of_universum;
-    RelativeSize upper_limit_rational = (relative_range_offset + relative_range_size) * size_of_universum;
+    RelativeSize lower_limit_rational = range_lower + relative_range_offset * size_of_universum;
+    RelativeSize upper_limit_rational = range_lower + (relative_range_offset + relative_range_size) * size_of_universum;
 
     UInt64 lower = boost::rational_cast<ASTSampleRatio::BigNum>(lower_limit_rational);
     UInt64 upper = boost::rational_cast<ASTSampleRatio::BigNum>(upper_limit_rational);
 
-    if (lower > 0)
+    if (lower_limit_rational > range_lower)
```

Comment:
```suggestion
    if (lower_limit_rational > 0)
```

the point is if it's 0, then you can read everything `<= upper_limit`

Reply:
I think this also relates to my comment below. `parallel_replicas_custom_key_range_lower` is not intended to actually filter out any values. So we need this line to know if we should include a condition for the lowest segment of the range. Also when `parallel_replicas_custom_key_range_lower = 0`, range_lower will also be 0.

Reply:
I see, makes sense
1 replica will catch all values `< lower` and 1 will catch all values `> upper`